### ISSUE TITLE
Operators' presedence seems incompatible with Runes' and Haskell's

### ISSUE THREAD

**Issue by koher** [tryswift/TryParsec#9]
[Precedence of some infix operators declared in TryParsec](https://github.com/tryswift/TryParsec/blob/8946a764d46d003045993d747ec631a49fd99449/Sources/TryParsec/Operators.swift) seems incompatible with ones [in thoughtbot/Runes](https://github.com/thoughtbot/Runes/blob/42fbd8fb52b8155e3eb4ebe8b82bcea1c8d8bff6/Source/Runes.swift) and in Haskell. Any reasons for this?

In Haskell:

- `<*>`: `4` https://hackage.haskell.org/package/base-4.8.2.0/docs/Control-Applicative.html#v:-60--42--62-
- `<$>`: `4` http://hackage.haskell.org/package/base-4.8.2.0/docs/Control-Applicative.html#v:-60--36--62-
- `<|>`: `3` http://hackage.haskell.org/package/base-4.8.2.0/docs/Control-Applicative.html#v:-60--124--62-

`<*>` and `<$>`'s precedence are equal to `==`'s, and `<|>`'s is equal to `&&`'s.

https://www.haskell.org/onlinereport/decls.html

In Swift, `==`'s precedence is `130` and `&&`'s is `120`.

https://developer.apple.com/library/tvos/documentation/Swift/Reference/Swift_StandardLibrary_Operators/index.html#//apple_ref/doc/uid/TP40016054

In TryParsec:

```swift
infix operator <|>  { associativity right precedence 130 }
infix operator <*>  { associativity left precedence 140 }
infix operator <^>  { associativity left precedence 140 }
```

In Runes:

```swift
infix operator <^> {
    associativity left

    // Same precedence as the equality operator (`==`)
    precedence 130
}

/**
apply a function with context to a value with context
Expected function type: `f (a -> b) -> f a -> f b`
*/
infix operator <*> {
    associativity left

    // Same precedence as the equality operator (`==`)
    precedence 130
}
```

**Comment by inamiy** [comment]
@koher
Thank you for detailed comment!
I did not use Runes because there were no Applicative operators, and decided to use [Swiftz v0.2.0](https://github.com/typelift/Swiftz/blob/v0.2.0/Swiftz/Operators.swift) instead. You are right that we should also consider precedences of `==` and `&&`.

By the way, it seems Swiftz team has introduced [Operadics](https://github.com/typelift/Operadics/blob/master/Operators.swift) nowadays, which we can think of using it instead. Both libraries still miss `<|>` operator (at least), and needs a patch in order to adopt either one as a new dependency.

**Comment by koher** [comment]
@inamiy 
How about using `??` instead of `<|>`?

I think using Runes or Operadics is not so important. The important thing is to have compatible precedences, or operators may collide. Because `<*>` and `<^>`'s precedences are `130` in both Runes and Operadics, it seems good that TryParsec follows it.

**Comment by inamiy** [comment]
@koher 
I opened pull-request in #11. Will you please check it?

By the way, I'm not positive of using `??` operator instead of `<|>` because:

- It has `precedence 131` in Swift core, which is slightly higher than `Haskell.<|>`'s `infixl 3` (`precedence 130` is already used for `infix 4` operators)
- Though function signature: 

   ```swift
   // overloaded one
   public func ??<T>(optional: T?, @autoclosure defaultValue: () throws -> T?) rethrows -> T?
   ``` 

   looks very alike to `<|>`, I personally think the above method is a special case for beginner Swift programmers to ease Optional handling, and `??` should mostly be treated as unwrap-or-default operation.

**Comment by inamiy** [comment]
Closed as #15 is merged.
